rules/aip0231/request_parent_field.go

```diff
@@ -16,17 +16,41 @@ package aip0231
 
 import (
 	"fmt"
+	"strings"
 
 	"github.com/googleapis/api-linter/lint"
 	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
 	"github.com/jhump/protoreflect/desc"
 	"github.com/jhump/protoreflect/desc/builder"
+	"github.com/stoewer/go-strcase"
 )
 
 // The Batch Get request message should have parent field.
-var parentField = &lint.MessageRule{
-	Name:   lint.NewRuleName(231, "request-parent-field"),
-	OnlyIf: isBatchGetRequestMessage,
+var requestParentField = &lint.MessageRule{
+	Name: lint.NewRuleName(231, "request-parent-field"),
+	OnlyIf: func(m *desc.MessageDescriptor) bool {
+		// Sanity check: If the resource has a pattern, and that pattern
+		// contains only one variable, then a parent field is not expected.
+		//
+		// In order to parse out the pattern, we get the resource message
+		// from the response, then get the resource annotation from that,
+		// and then inspect the pattern there (oy!).
+		plural := strings.TrimPrefix(strings.TrimSuffix(m.GetName(), "Request"), "BatchGet")
+		if resp := m.GetFile().FindMessage(fmt.Sprintf("BatchGet%sResponse", plural)); resp != nil {
+			if paged := resp.FindFieldByName(strcase.SnakeCase(plural)); paged != nil {
+				if resource := utils.GetResource(paged.GetMessageType()); resource != nil {
+					for _, pattern := range resource.GetPattern() {
+						if strings.Count(pattern, "{") == 1 {
+							return false
+						}
+					}
+				}
+			}
+		}
```

Comment:
Can this logic (which is used for BatchGet, BatchCreate, and BatchUpdate) be pulled out into a function that takes the "batching method type" as param i.e. `"BatchGet"`, as well as the necessary proto input i.e. `m *desc.MessageDescriptor`? This code is non-trivial and used in three places, feels ripe for a `func`. WDYT? If there is a nuance in the code for each that prevents this that I missed, lmk.

Reply:
It _can_ be, although the code is slightly different each time, and I am not sure there is a good place to put it. I do not know that it is common enough to justify going in `utils`. This is right in the sour spot where the function can not be in the AIP modules themselves, but are not generic enough for me to be convinced that they belong in utils.

Reply:
Ok, I understand. Bummer!